pkg/kocache/kocache.go

```diff
@@ -59,34 +72,70 @@ func (o *Options) applyDefaults() {
 	if o.CleanupInterval == 0 {
 		o.CleanupInterval = cleanupInterval
 	}
+	o.StartOnline = true
+	return o
 }
 
 type koCache struct {
-	opts Options
+	opts *options
 
+	parentCtx    context.Context
 	entries      map[string]*entry
-	entriesMutex sync.Mutex
+	entriesMutex *sync.Mutex
 
-	upstreamClient  *http.Client
+	upstreamClient  httpClient
 	upstreamBaseURL string
+
+	centralReady *atomic.Bool
+	// onlineCtx will be canceled when connectivity to central is lost
+	onlineCtx       context.Context
+	onlineCtxCancel context.CancelCauseFunc
+	ctxMutex        *sync.Mutex
+}
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
 }
 
 // New returns a new kernel object cache whose lifetime is bound by the given context, using the given
 // HTTP client and base URL for upstream requests.
-func New(ctx context.Context, upstreamClient *http.Client, upstreamBaseURL string, opts Options) probeupload.ProbeSource {
-	opts.applyDefaults()
+func New(parentCtx context.Context, upstreamClient httpClient, upstreamBaseURL string, opts ...func(o *options)) *koCache {
+	opt := applyDefaults(&options{})
+	for _, option := range opts {
+		option(opt)
+	}
+	onlineCtx, onlineCtxCancel := context.WithCancelCause(parentCtx)
 	cache := &koCache{
-		opts:            opts,
+		opts:            opt,
+		parentCtx:       parentCtx,
 		entries:         make(map[string]*entry),
+		entriesMutex:    &sync.Mutex{},
 		upstreamClient:  upstreamClient,
 		upstreamBaseURL: strings.TrimSuffix(upstreamBaseURL, "/"),
+		centralReady:    &atomic.Bool{},
+		onlineCtx:       onlineCtx,
+		onlineCtxCancel: onlineCtxCancel,
+		ctxMutex:        &sync.Mutex{},
 	}
+	cache.centralReady.Store(opt.StartOnline)
```

Comment:
If you start offline, shouldn't the context be already canceled? Otherwise it seems like there are 3 possible states the component can be in:

1. `centralReady: true` + `context: not canceled` -> Started online or `GoOnline` was called
2. `centralReady: false` + `context: canceled` -> `GoOffline` was called
3. `centralReady: false` + `context: not canceled` -> Started offline

Not sure if this significantly changes the behavior of the component. But it would be nice not have state (3) as a possibility and make sure the initial context is canceled here.

Reply:
You are right, the atomic bool and the context should be in sync. I think it would work the way it is currently, but it is ugly and may be confusing. Let me fix that.

Reply:
I removed the atomic bool and added the `offlineCtr` object to control the state. WDYT?